src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java

```diff
@@ -19,7 +23,33 @@
 public abstract class AMemoryOperation extends AMemoryAsset implements Operation {
 
     protected AMemoryOperation(String metaString, MemoryAgent memoryAgent) {
-        super(metaString, memoryAgent);
+        super(getMetaWithProv(metaString,memoryAgent), memoryAgent);
+    }
+
+    /**
+     * Method to add provenace in the metadata if not present for all Memory Operation
+     * @param metaString Metadata as string
+     * @param memoryAgent agent
+     * @return metadata with PROVENANCE
+     */
+    private static String getMetaWithProv(String metaString,MemoryAgent memoryAgent) {
+
+        Map<String,Object> meta = JSON.toMap(metaString);
+        //check if already provenance is present
+
+        if(meta.get(Constant.PROVENANCE)== null) {
+            // adding default provenance
+            String actId = UUID.randomUUID().toString();
+            // check if DID is present in agent then agent id will be did
+            // else will create a random UUID
+            String agentId = memoryAgent.getDID()!=null?
+                    memoryAgent.getDID().toString():
+                    UUID.randomUUID().toString();
```

Comment:
Definitely should not be creating random UUIDs here!

@shark8me please ensure that the provenance DEP specifies what happens by default (probably a "should" or "may")

Reply:
[DEP12](https://github.com/DEX-Company/DEPs/tree/master/12#agents-1) does specify what the agent id should be